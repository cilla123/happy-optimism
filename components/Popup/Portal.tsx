import React from 'react';
import ReactDOM from 'react-dom';

const createPortal = (ReactDOM as any).createPortal;

export interface PortalProps {
  prefixCls?: string;
  className?: any;
  getContainer: Function;
}
export default class Portal extends React.Component<PortalProps, any> {
  _container: any;

  constructor(props) {
    super(props);
    this._container = this.props.getContainer();
  }

  // shouldComponentUpdate() {
  //   return false;
  // }

  render() {
    if (this.props.children) {
      return createPortal(this.props.children, this._container);
    }
    return null as any;
  }
}
