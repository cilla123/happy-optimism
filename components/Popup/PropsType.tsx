export default interface PropsType {
  visible?: boolean;
  direction?: 'top' | 'right' | 'bottom' | 'left';
  autoClose?: boolean;
  stayTime?: number;
  animationDuration?: number;
  onClose?: () => void;
  mask?: boolean;
  maskType?: 'transparent' | 'normal';
  onMaskClick?: () => void;
}
