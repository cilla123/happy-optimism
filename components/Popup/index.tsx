import React, { PureComponent } from 'react';
import ReactDOM from 'react-dom';
import classnames from 'classnames';
import PropsType from './PropsType';
import Events from '../utils/events';
import Mask from '../Mask';
import Portal from './Portal';

const IS_REACT_16 = !!(ReactDOM as any).createPortal;
export interface PopupProps extends PropsType {
  prefixCls?: string;
  className?: string;
}

export default class Popup extends PureComponent<PopupProps, any> {

  static defaultProps = {
    prefixCls: 'za-popup',
    visible: false,
    mask: true,
    direction: 'bottom',
    autoClose: false,
    stayTime: 3000,
    animationDuration: 200,
    maskType: Mask.defaultProps.type,
  };

  private timer: number;
  private container;
  private popup;

  constructor(props) {
    super(props);
    this.state = {
      isShow: props.visible || false,
      isMaskShow: props.visible || false,
      isPending: false,
      animationState: 'enter',
    };
  }

  componentDidMount() {
    this.renderPopup();
    Events.on(this.popup, 'webkitTransitionEnd', this.animationEnd);
    Events.on(this.popup, 'transitionend', this.animationEnd);
  }

  componentWillReceiveProps(nextProps) {
    clearTimeout(this.timer);

    if (nextProps.visible) {
      this.enter(nextProps);
    } else {
      this.leave();
    }
  }

  componentWillUnmount() {
    if (!IS_REACT_16) {
      ReactDOM.unmountComponentAtNode(this.container);
    }
    document.body.removeChild(this.container);
    Events.off(this.popup, 'webkitTransitionEnd', this.animationEnd);
    Events.off(this.popup, 'transitionend', this.animationEnd);
  }

  renderPopup() {
    if (IS_REACT_16) {
      return;
    }
    ReactDOM.unstable_renderSubtreeIntoContainer(
      this,
      this.getComponent(),
      this.getContainer(),
    );
  }

  enter = ({ stayTime, autoClose, onMaskClick }) => {
    this.setState({
      isShow: true,
      isMaskShow: true,
      isPending: true,
      animationState: 'enter',
    });

    if (stayTime > 0 && autoClose) {
      this.timer = setTimeout(() => {
        onMaskClick();
        clearTimeout(this.timer);
      }, stayTime);
    }
  }

  leave = () => {
    this.setState({
      isShow: false,
      isPending: true,
      animationState: 'leave',
    });
  }

  animationEnd = () => {
    const { onClose } = this.props;
    const { animationState } = this.state;

    if (animationState === 'leave') {
      this.setState({
        isMaskShow: false,
      });
      if (typeof onClose === 'function') {
        onClose();
      }
    }
  }

  renderMask = () => {
    const { mask, maskType, onMaskClick, animationDuration } = this.props;
    const { isPending, animationState, isMaskShow } = this.state;

    const maskCls = classnames({
      [`fade-${animationState}`]: isPending,
    });

    const maskStyle = {
      WebkitAnimationDuration: `${animationDuration}ms`,
      animationDuration: `${animationDuration}ms`,
    };

    return mask && (
      <Mask
        className={maskCls}
        style={maskStyle}
        visible={isMaskShow}
        type={maskType}
        onClose={onMaskClick}
      />
    );
  }

  getContainer() {
    let container = document.querySelector(`#${this.props.prefixCls}-container`);
    if (!container) {
      container = document.createElement('div');
      container.classList.add('popup-container');
      document.body.appendChild(container);
    }
    this.container = container;
    return container;
  }

  getComponent() {
    const { prefixCls, className, animationDuration, direction, children } = this.props;
    const { isShow } = this.state;

    const popupCls = classnames(`${prefixCls}`, className, {
      [`${prefixCls}-hidden`]: !isShow,
    });
    const wrapCls = classnames(`${prefixCls}-wrapper`, `${prefixCls}-wrapper-${direction}`);

    const wrapStyle = {
      WebkitTransitionDuration: `${animationDuration}ms`,
      transitionDuration: `${animationDuration}ms`,
    };
    return (
      <div className={popupCls} ref={(popup) => { this.popup = popup; }}>
        <div className={wrapCls} style={wrapStyle}>
          {children}
        </div>
        {this.renderMask()}
      </div>
    );
  }

  renderPortal() {
    if (!IS_REACT_16) {
      return null;
    }

    const portal = (
      <Portal getContainer={() => this.getContainer()}>
        {this.getComponent()}
      </Portal>
    );

    return portal;
  }

  render() {
    return this.renderPortal();
  }
}
