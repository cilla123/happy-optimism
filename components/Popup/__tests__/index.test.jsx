import React from 'react';
import ReactDOM from 'react-dom';
import { render, shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Popup from '../index';

describe('Popup', () => {
  it('renders correctly', () => {
    const onMaskClick = jest.fn();
    const onClose = jest.fn();
    const wrapper = mount(
      <Popup
        direction="bottom"
        onMaskClick={onMaskClick}
        onClose={onClose}>
        foo
      </Popup>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('visible change false', () => {
    jest.useFakeTimers();
    const wrapper = mount(
      <Popup>foo</Popup>
    );
    wrapper.setProps({ visible: false });
    jest.runAllTimers();
    wrapper.unmount();
  });

  // it('stayTime is 0', () => {
  //   const wrapper = shallow(
  //     <Popup stayTime={0}>foo</Popup>
  //   );
  //   wrapper.setProps({ visible: true });
  // });

  it('autoClose', () => {
    jest.useFakeTimers();
    const onMaskClick = jest.fn();
    const onClose = jest.fn();
    const wrapper = mount(
      <Popup
        visible
        autoClose
        onMaskClick={onMaskClick}
        onClose={onClose}>foo</Popup>
    );
    wrapper.setProps({ visible: true });
    jest.runAllTimers();
  });
});
