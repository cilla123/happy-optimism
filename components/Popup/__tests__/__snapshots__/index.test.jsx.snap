// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup renders correctly 1`] = `
<Popup
  animationDuration={200}
  autoClose={false}
  direction="bottom"
  mask={true}
  maskType="normal"
  onClose={[Function]}
  onMaskClick={[Function]}
  prefixCls="za-popup"
  stayTime={3000}
  visible={false}
>
  <div>
    <Portal
      getContainer={[Function]}
    >
      <div
        className="za-popup za-popup-hidden"
      >
        <div
          className="za-popup-wrapper za-popup-wrapper-bottom"
          style={
            Object {
              "WebkitTransitionDuration": "200ms",
              "transitionDuration": "200ms",
            }
          }
        >
          foo
        </div>
        <Mask
          className=""
          onClose={[Function]}
          prefixCls="za-mask"
          style={
            Object {
              "WebkitAnimationDuration": "200ms",
              "animationDuration": "200ms",
            }
          }
          type="normal"
          visible={false}
        />
      </div>
    </Portal>
  </div>
</Popup>
`;
